source/Nevermore/Util/DataModificationQueryBuilder.cs

```diff
@@ -330,87 +329,155 @@ CommandParameterValues GetDocumentParameters(Func<DocumentMap, string> allocateI
             return result;
         }
 
-
-        void AppendRelatedDocumentStatementsForInsert(
-            StringBuilder sb,
-            CommandParameterValues parameters,
+        PreparedCommand[] CreateInsertCommandsWithRelatedDocuments(
+            StringBuilder insertStatementBuilder,
             DocumentMap mapping,
-            IReadOnlyList<object> documents)
+            IReadOnlyList<object> documents,
+            InsertOptions options)
         {
+            var commands = new List<PreparedCommand>();
+            var parameters = new CommandParameterValues();
+            // we need to create these document parameters for each command - parameters are disposed at the end of execution in 'CommandExecutor'
+            Func<CommandParameterValues> documentParametersCreator = () => GetDocumentParameters(m => keyAllocator(m), options.CustomAssignedId, documents, mapping, DataModification.Insert);
+            parameters.AddRange(documentParametersCreator());
+
             var relatedDocumentData = GetRelatedDocumentTableData(mapping, documents);
+            if (!relatedDocumentData.Any(x => x.Related.Any()))
+            {
+                commands.Add(new PreparedCommand(insertStatementBuilder.ToString(), parameters, RetriableOperation.Insert, mapping, options.CommandTimeout));
+                return commands.ToArray();
+            }
 
             foreach (var data in relatedDocumentData.Where(g => g.Related.Length > 0))
             {
                 var relatedVariablePrefix = $"{data.TableName.ToLower()}_";
 
-                sb.AppendLine($"INSERT INTO [{data.SchemaName}].[{data.TableName}] ([{data.IdColumnName}], [{data.IdTableColumnName}], [{data.RelatedDocumentIdColumnName}], [{data.RelatedDocumentTableColumnName}]) VALUES");
-                var related = data.Related;
-
-                for (var x = 0; x < related.Length; x++)
+                var remaining = data.Related.Length;
+                while (remaining > 0)
```

Comment:
I agree with Frank here. The while loop is hard to understand. What about this high level approach:

`relatedDocumentToInsert.SplitIntoBatches((FirstBatchSize, commandWithExistingInsertSQL), (OtherBatchSizes, new Command())).Select((command, batch) => add batch to the command).ToListOfCommands()`.  I would leave this one until the very end as it is just the refactor.

Reply:
@pawelpabich there's a bug with INSERTS and UPDATES that doesn't handle multiple related documents. I've pushed up a commit to expose this: https://github.com/OctopusDeploy/Nevermore/pull/153/commits/5ad88f8a0c21bef614bb995f7d27aaa1177a6670